File: expense-tracker/app/src/main/java/com/farhan/paisatrack/ui/screens/SettingsScreen.kt
package com.farhan.paisatrack.ui.screens

import android.Manifest
import android.os.Build
import android.widget.Toast
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.wrapContentSize
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowForward
import androidx.compose.material.icons.filled.Category
import androidx.compose.material.icons.filled.DarkMode
import androidx.compose.material.icons.filled.Download
import androidx.compose.material.icons.filled.Notifications
import androidx.compose.material.icons.filled.Sms
import androidx.compose.material.icons.filled.Wallet
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TimePicker
import androidx.compose.material3.rememberTimePickerState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.farhan.paisatrack.ui.MainViewModel
import com.farhan.paisatrack.ui.components.IconBadge
import com.farhan.paisatrack.ui.components.LabeledDropdown
import com.farhan.paisatrack.ui.components.SectionCard
import com.farhan.paisatrack.ui.theme.Amber
import com.farhan.paisatrack.ui.theme.Mint
import com.farhan.paisatrack.ui.theme.Sky
import com.farhan.paisatrack.ui.theme.Violet

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun SettingsScreen(
    vm: MainViewModel,
    isDark: Boolean,
    onToggleTheme: () -> Unit,
    onManageAccounts: () -> Unit,
    onManageCategories: () -> Unit
) {
    val context = LocalContext.current
    val accounts by vm.accounts.collectAsStateWithLifecycle()

    var reminderOn by remember { mutableStateOf(vm.prefs.reminderEnabled) }
    var reminderHour by remember { mutableIntStateOf(vm.prefs.reminderHour) }
    var reminderMinute by remember { mutableIntStateOf(vm.prefs.reminderMinute) }
    var smsOn by remember { mutableStateOf(vm.prefs.smsCaptureEnabled) }
    var defaultUpi by remember { mutableStateOf(vm.prefs.defaultUpiAccountId) }
    var showTimeDialog by remember { mutableStateOf(false) }

    val notifPermLauncher = rememberLauncherForActivityResult(
        ActivityResultContracts.RequestPermission()
    ) { }
    val smsPermLauncher = rememberLauncherForActivityResult(
        ActivityResultContracts.RequestMultiplePermissions()
    ) { result ->
        val granted = result.values.all { it }
        if (granted) {
            vm.importSmsInbox { added ->
                Toast.makeText(context, "Imported $added transactions from SMS", Toast.LENGTH_LONG).show()
            }
        } else {
            Toast.makeText(context, "SMS permission needed to read transactions", Toast.LENGTH_SHORT).show()
        }
    }

    Column(
        Modifier
            .fillMaxWidth()
            .verticalScroll(rememberScrollState())
            .padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        Text("Settings", style = MaterialTheme.typography.headlineSmall, fontWeight = FontWeight.Bold)

        // Daily reminder
        SectionCard {
            Row(verticalAlignment = Alignment.CenterVertically) {
                IconBadge(Icons.Filled.Notifications, Violet)
                Spacer(Modifier.width(12.dp))
                Column(Modifier.weight(1f)) {
                    Text("Daily reminder", fontWeight = FontWeight.SemiBold)
                    Text("Nudge me to log expenses", fontSize = 12.sp, color = MaterialTheme.colorScheme.onSurfaceVariant)
                }
                Switch(checked = reminderOn, onCheckedChange = {
                    reminderOn = it
                    vm.setReminder(it, reminderHour, reminderMinute)
                    if (it && Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                        notifPermLauncher.launch(Manifest.permission.POST_NOTIFICATIONS)
                    }
                })
            }
            if (reminderOn) {
                Spacer(Modifier.height(8.dp))
                Row(
                    Modifier.fillMaxWidth().clickable { showTimeDialog = true },
                    horizontalArrangement = Arrangement.SpaceBetween
                ) {
                    Text("Reminder time")
                    Text(
                        "%02d:%02d".format(reminderHour, reminderMinute),
                        fontWeight = FontWeight.SemiBold,
                        color = MaterialTheme.colorScheme.primary
                    )
                }
            }
        }

        // SMS capture
        SectionCard {
            Row(verticalAlignment = Alignment.CenterVertically) {
                IconBadge(Icons.Filled.Sms, Mint)
                Spacer(Modifier.width(12.dp))
                Column(Modifier.weight(1f)) {
                    Text("Auto-capture from SMS", fontWeight = FontWeight.SemiBold)
                    Text("Detect UPI / bank transactions automatically", fontSize = 12.sp, color = MaterialTheme.colorScheme.onSurfaceVariant)
                }
                Switch(checked = smsOn, onCheckedChange = {
                    smsOn = it
                    vm.setSmsCapture(it)
                    if (it) {
                        smsPermLauncherRequestOnly(smsPermLauncher)
                    }
                })
            }
            Spacer(Modifier.height(10.dp))
            Row(
                Modifier.fillMaxWidth().clickable {
                    smsPermLauncher.launch(arrayOf(Manifest.permission.READ_SMS, Manifest.permission.RECEIVE_SMS))
                },
                verticalAlignment = Alignment.CenterVertically
            ) {
                Icon(Icons.Filled.Download, contentDescription = null, tint = MaterialTheme.colorScheme.primary)
                Spacer(Modifier.width(8.dp))
                Text("Scan SMS inbox now", color = MaterialTheme.colorScheme.primary, fontWeight = FontWeight.SemiBold)
            }
        }

        // Default UPI account
        SectionCard {
            Row(verticalAlignment = Alignment.CenterVertically) {
                IconBadge(Icons.Filled.Wallet, Sky)
                Spacer(Modifier.width(12.dp))
                Text("Default account for detected UPI", fontWeight = FontWeight.SemiBold, modifier = Modifier.weight(1f))
            }
            Spacer(Modifier.height(10.dp))
            LabeledDropdown(
                label = "Account",
                options = accounts,
                selected = accounts.firstOrNull { it.id == defaultUpi },
                optionLabel = { it.name },
                onSelect = {
                    defaultUpi = it.id
                    vm.setDefaultUpiAccount(it.id)
                }
            )
        }

        // Manage
        NavRow("Manage accounts", "Add or edit your banks, cash & cards", Icons.Filled.Wallet, Amber, onManageAccounts)
        NavRow("Manage categories", "Customize your expense & income types", Icons.Filled.Category, Violet, onManageCategories)

        // Theme
        SectionCard {
            Row(verticalAlignment = Alignment.CenterVertically) {
                IconBadge(Icons.Filled.DarkMode, MaterialTheme.colorScheme.onSurface)
                Spacer(Modifier.width(12.dp))
                Text("Dark theme", fontWeight = FontWeight.SemiBold, modifier = Modifier.weight(1f))
                Switch(checked = isDark, onCheckedChange = { onToggleTheme() })
            }
        }

        SectionCard {
            Text("🔒 100% on your device", fontWeight = FontWeight.SemiBold)
            Spacer(Modifier.height(4.dp))
            Text(
                "All your data is stored locally in this app. Nothing is uploaded to any server. " +
                    "If you uninstall the app, your data is removed — keep a device backup if needed.",
                fontSize = 13.sp,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }
        Spacer(Modifier.height(80.dp))
    }

    if (showTimeDialog) {
        val state = rememberTimePickerState(initialHour = reminderHour, initialMinute = reminderMinute, is24Hour = true)
        AlertDialog(
            onDismissRequest = { showTimeDialog = false },
            confirmButton = {
                TextButton(onClick = {
                    reminderHour = state.hour
                    reminderMinute = state.minute
                    vm.setReminder(reminderOn, reminderHour, reminderMinute)
                    showTimeDialog = false
                }) { Text("OK") }
            },
            dismissButton = { TextButton(onClick = { showTimeDialog = false }) { Text("Cancel") } },
            text = { Box(Modifier.wrapContentSize()) { TimePicker(state = state) } }
        )
    }
}

private fun smsPermLauncherRequestOnly(
    launcher: androidx.activity.result.ActivityResultLauncher<Array<String>>
) {
    launcher.launch(arrayOf(Manifest.permission.READ_SMS, Manifest.permission.RECEIVE_SMS))
}

@Composable
private fun NavRow(
    title: String,
    subtitle: String,
    icon: androidx.compose.ui.graphics.vector.ImageVector,
    color: androidx.compose.ui.graphics.Color,
    onClick: () -> Unit
) {
    SectionCard(modifier = Modifier.clickable { onClick() }) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            IconBadge(icon, color)
            Spacer(Modifier.width(12.dp))
            Column(Modifier.weight(1f)) {
                Text(title, fontWeight = FontWeight.SemiBold)
                Text(subtitle, fontSize = 12.sp, color = MaterialTheme.colorScheme.onSurfaceVariant)
            }
            Icon(Icons.AutoMirrored.Filled.ArrowForward, contentDescription = null, tint = MaterialTheme.colorScheme.onSurfaceVariant)
        }
    }
}

File: farhans-wallet-flutter/lib/features/settings/settings_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../core/theme/tokens.dart';
import '../../core/widgets/bits.dart';
import '../../core/widgets/buttons.dart';
import '../../core/widgets/detail_scaffold.dart';
import '../../core/widgets/fields.dart';
import '../../core/widgets/glass.dart';
import '../../state/wallet_state.dart';
import 'accounts_screen.dart';
import 'categories_screen.dart';

class SettingsScreen extends ConsumerStatefulWidget {
  const SettingsScreen({super.key});

  @override
  ConsumerState<SettingsScreen> createState() => _SettingsScreenState();
}

class _SettingsScreenState extends ConsumerState<SettingsScreen> {
  @override
  Widget build(BuildContext context) {
    final t = context.tokens;
    final prefs = ref.read(prefsProvider);

    return DetailScaffold(
      title: 'Settings',
      child: ListView(
        padding: const EdgeInsets.fromLTRB(Insets.lg, Insets.xs, Insets.lg, Insets.xl),
        children: [
          // Daily reminder
          GlassCard(
            child: Column(
              children: [
                _row(
                  Icons.notifications_active_rounded,
                  t.accentA,
                  'Daily reminder',
                  'Nudge me every night to log expenses',
                  Switch(
                    value: prefs.reminderEnabled,
                    activeThumbColor: t.accentA,
                    onChanged: (v) async {
                      setState(() => prefs.reminderEnabled = v);
                      final notif = ref.read(notificationProvider);
                      if (v) {
                        await notif.requestPermission();
                        await notif.scheduleDaily(prefs.reminderHour, prefs.reminderMinute);
                      } else {
                        await notif.cancelReminder();
                      }
                    },
                  ),
                ),
                if (prefs.reminderEnabled) ...[
                  const SizedBox(height: Insets.xs),
                  Pressable(
                    onTap: () async {
                      final picked = await showTimePicker(
                        context: context,
                        initialTime: TimeOfDay(hour: prefs.reminderHour, minute: prefs.reminderMinute),
                      );
                      if (picked != null) {
                        setState(() {
                          prefs.reminderHour = picked.hour;
                          prefs.reminderMinute = picked.minute;
                        });
                        await ref.read(notificationProvider).scheduleDaily(picked.hour, picked.minute);
                      }
                    },
                    child: Row(
                      mainAxisAlignment: MainAxisAlignment.spaceBetween,
                      children: [
                        Text('Reminder time', style: context.text.bodyLarge),
                        Text(
                          '${prefs.reminderHour.toString().padLeft(2, '0')}:${prefs.reminderMinute.toString().padLeft(2, '0')}',
                          style: TextStyle(color: t.accentA, fontWeight: FontWeight.w700),
                        ),
                      ],
                    ),
                  ),
                ],
              ],
            ),
          ),
          const SizedBox(height: Insets.sm),
          // SMS capture
          GlassCard(
            child: Column(
              children: [
                _row(
                  Icons.sms_rounded,
                  t.success,
                  'Auto-capture from SMS',
                  'Transactions are added automatically when a bank/UPI SMS arrives',
                  Switch(
                    value: prefs.smsCapture,
                    activeThumbColor: t.accentA,
                    onChanged: (v) => setState(() => prefs.smsCapture = v),
                  ),
                ),
                const SizedBox(height: Insets.sm),
                Pressable(
                  onTap: () async {
                    final sms = ref.read(smsServiceProvider);
                    final granted = await sms.requestPermission();
                    if (!granted) {
                      if (!context.mounted) return;
                      ScaffoldMessenger.of(context).showSnackBar(const SnackBar(
                          content: Text('SMS permission is needed to read transactions')));
                      return;
                    }
                    await ref.read(walletProvider.notifier).scanNewSms();
                    if (!context.mounted) return;
                    ScaffoldMessenger.of(context).showSnackBar(
                        const SnackBar(content: Text('Checked for new transactions')));
                  },
                  child: Row(
                    children: [
                      Icon(Icons.refresh_rounded, color: t.accentA),
                      const SizedBox(width: Insets.xs),
                      Text('Check for new SMS now',
                          style: TextStyle(color: t.accentA, fontWeight: FontWeight.w600)),
                    ],
                  ),
                ),
                const SizedBox(height: Insets.xs),
                Text(
                  'New SMS are added automatically (debit or credit). The account is set from '
                  'the bank name in the SMS (e.g. name an account “Kotak”); otherwise it’s left '
                  'unset for you to pick. Only messages received after installing are captured.',
                  style: context.text.bodySmall?.copyWith(color: t.textMid),
                ),
              ],
            ),
          ),
          const SizedBox(height: Insets.sm),
          _navCard(Icons.account_balance_rounded, t.warning, 'Manage accounts',
              'Banks, cash & cards', () {
            Navigator.push(context, MaterialPageRoute(builder: (_) => const AccountsScreen()));
          }),
          const SizedBox(height: Insets.sm),
          _navCard(Icons.category_rounded, t.accentA, 'Manage categories',
              'Customize expense & income types', () {
            Navigator.push(context, MaterialPageRoute(builder: (_) => const CategoriesScreen()));
          }),
          const SizedBox(height: Insets.sm),
          // Theme
          GlassCard(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                _row(Icons.palette_rounded, t.textHigh, 'Appearance', null, null),
                const SizedBox(height: Insets.sm),
                AppSelectField<ThemeMode>(
                  label: 'Theme',
                  value: ref.watch(themeModeProvider),
                  options: const [ThemeMode.system, ThemeMode.light, ThemeMode.dark],
                  optionLabel: (m) => switch (m) {
                    ThemeMode.system => 'System',
                    ThemeMode.light => 'Light',
                    ThemeMode.dark => 'Dark',
                  },
                  onChanged: (m) {
                    ref.read(themeModeProvider.notifier).state = m;
                    prefs.themeMode = switch (m) {
                      ThemeMode.light => 1,
                      ThemeMode.system => 0,
                      ThemeMode.dark => 2,
                    };
                  },
                ),
              ],
            ),
          ),
          const SizedBox(height: Insets.sm),
          // Backup & restore
          GlassCard(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                _row(Icons.cloud_sync_rounded, t.info, 'Backup & restore', null, null),
                const SizedBox(height: Insets.sm),
                Container(
                  padding: const EdgeInsets.all(Insets.sm),
                  decoration: BoxDecoration(
                    color: t.glassFill,
                    borderRadius: BorderRadius.circular(Corners.sm),
                    border: Border.all(color: t.glassBorder),
                  ),
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text('Move your data to another phone', style: context.text.titleSmall),
                      const SizedBox(height: 6),
                      _step(context, '1', 'Tap Export & share below and save the backup file (to Drive, WhatsApp, Files…).'),
                      _step(context, '2', 'Install Pocket Flow on the other phone.'),
                      _step(context, '3', 'There, open Settings → Backup & restore → Import and pick that file.'),
                      const SizedBox(height: 6),
                      Text(
                        'Tip: export now and then so a backup exists before you ever uninstall — '
                        'the app also opts into Android auto-backup, but an exported file is the sure way.',
                        style: context.text.bodySmall?.copyWith(color: t.textMid),
                      ),
                    ],
                  ),
                ),
                const SizedBox(height: Insets.md),
                Row(
                  children: [
                    Expanded(
                      child: PrimaryButton(
                        label: 'Export',
                        icon: Icons.ios_share_rounded,
                        onPressed: () async {
                          final data = await ref.read(walletProvider.notifier).exportData();
                          await ref.read(backupServiceProvider).exportAndShare(data);
                        },
                      ),
                    ),
                    const SizedBox(width: Insets.sm),
                    Expanded(
                      child: SecondaryButton(
                        label: 'Import',
                        icon: Icons.download_rounded,
                        onPressed: () => _import(context),
                      ),
                    ),
                  ],
                ),
              ],
            ),
          ),
          const SizedBox(height: Insets.sm),
          GlassCard(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text('🔒 100% on your device', style: context.text.titleSmall),
                const SizedBox(height: 4),
                Text(
                  'All your data is stored locally — nothing is uploaded to any server. '
                  'Uninstalling removes the local copy, so keep an exported backup if you want to restore later.',
                  style: context.text.bodyMedium?.copyWith(color: t.textMid),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }

  Future<void> _import(BuildContext context) async {
    final data = await ref.read(backupServiceProvider).pickAndRead();
    if (data == null || !context.mounted) return;
    final ok = await showDialog<bool>(
      context: context,
      builder: (ctx) => AlertDialog(
        backgroundColor: context.tokens.bgBottom,
        title: const Text('Restore from backup?'),
        content: const Text(
            'This replaces all current data in Pocket Flow with the contents of the backup file. Continue?'),
        actions: [
          TextButton(onPressed: () => Navigator.pop(ctx, false), child: const Text('Cancel')),
          TextButton(onPressed: () => Navigator.pop(ctx, true), child: const Text('Restore')),
        ],
      ),
    );
    if (ok != true || !context.mounted) return;
    await ref.read(walletProvider.notifier).importData(data);
    if (!context.mounted) return;
    ScaffoldMessenger.of(context)
        .showSnackBar(const SnackBar(content: Text('Data restored successfully')));
  }

  Widget _step(BuildContext context, String n, String text) {
    final t = context.tokens;
    return Padding(
      padding: const EdgeInsets.only(top: 6),
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Container(
            width: 20,
            height: 20,
            alignment: Alignment.center,
            decoration: BoxDecoration(color: t.accentA.withValues(alpha: 0.18), shape: BoxShape.circle),
            child: Text(n, style: TextStyle(color: t.accentA, fontSize: 11, fontWeight: FontWeight.w700)),
          ),
          const SizedBox(width: Insets.xs),
          Expanded(child: Text(text, style: context.text.bodySmall?.copyWith(color: t.textMid))),
        ],
      ),
    );
  }

  Widget _row(IconData icon, Color color, String title, String? subtitle, Widget? trailing) {
    final t = context.tokens;
    return Row(
      children: [
        IconBadge(icon: icon, color: color),
        const SizedBox(width: Insets.sm),
        Expanded(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Text(title, style: context.text.titleSmall),
              if (subtitle != null)
                Text(subtitle, style: context.text.bodySmall?.copyWith(color: t.textMid)),
            ],
          ),
        ),
        if (trailing != null) trailing,
      ],
    );
  }

  Widget _navCard(IconData icon, Color color, String title, String subtitle, VoidCallback onTap) {
    final t = context.tokens;
    return GlassCard(
      onTap: onTap,
      child: Row(
        children: [
          IconBadge(icon: icon, color: color),
          const SizedBox(width: Insets.sm),
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(title, style: context.text.titleSmall),
                Text(subtitle, style: context.text.bodySmall?.copyWith(color: t.textMid)),
              ],
            ),
          ),
          Icon(Icons.chevron_right_rounded, color: t.textMid),
        ],
      ),
    );
  }
}
